### PR TITLE
Some OSSA related optimization improvements

### PR DESCRIPTION
**Pull request by eeckstein**
Several (relatively) small optimization improvements to better deal with OSSA:

* GlobalPropertyOpt: handle `load_borrow` and `destroy_addr`
* COWArrayOpt: handle `load_borrow`
* LoopRotate: handle `copy_value` and `begin_borrow` correctly
* LoopRotate: don't rotate a loop if the new header is loop exiting as well
* LICM: (limited) support for OSSA
* SimplifyDestructure: canonicalize `destructure_tuple` and `destructure_struct` for trivial tuples/structs
* simplify `load_borrow`
* remove `begin_borrow` if the borrowed value is a `thin_to_thick_function`
* run `begin_borrow` simplification in SILCombine

For details see the commit messages.

This is part of fixing benchmark regressions when enabling OSSA modules.
rdar://139773406

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci benchmark

File: SwiftCompilerSources/Sources/Optimizer/InstructionSimplification/SimplifyBeginAndLoadBorrow.swift [hunk 1]
```diff
@@ -21,7 +21,52 @@ extension BeginBorrowInst : OnoneSimplifyable {
        !findPointerEscapingUse(of: borrowedValue)
     {
       tryReplaceBorrowWithOwnedOperand(beginBorrow: self, context)
+
+    } else if let thin2thickFn = borrowedValue as? ThinToThickFunctionInst,
+              uses.filterUsers(ofType: BranchInst.self).isEmpty
```
Comment (meg-gupta) [inline SwiftCompilerSources/Sources/Optimizer/InstructionSimplification/SimplifyBeginAndLoadBorrow.swift:26]: Looks like `ThinToThickFunctionInst` always produces values of none ownership. I am not sure if the borrowedValue can ever be `ThinToThickFunctionInst `. Can you add a test and move this to a separate commit if this is intended ?

**Comment by meg-gupta** [comment]
Awesome! I left a minor comment but this LGTM!

File: lib/SILOptimizer/LoopTransforms/LICM.cpp [hunk 2]
```diff
@@ -887,6 +902,10 @@ void LoopTreeOptimization::analyzeCurrentLoop(
   for (auto *BB : Loop->getBlocks()) {
     SmallVector<SILInstruction *, 8> sideEffectsInBlock;
     for (auto &Inst : *BB) {
+      if (hasOwnershipOperandsOrResults(&Inst)) {
```
Comment (meg-gupta) [inline lib/SILOptimizer/LoopTransforms/LICM.cpp:905]: Can you ensure there are tests covering other ownership instructions without side-effects ? load_borrow/destructure_struct/destructure_tuple etc

Comment (eeckstein) [reply lib/SILOptimizer/LoopTransforms/LICM.cpp:905]: I added a few tests - for instructions which really don't work if hoisted out of the loop.
This check is currently overly conservative. E.g. it is possible to hoist guaranteed values out of the loop.
Let's improve that when we convert the whole LICM optimization to OSSA.

**Comment by eeckstein** [comment]
@swift-ci test
